scripts/find-documents-missing-metadata.ts

```diff
@@ -0,0 +1,152 @@
+documentSnapshots.forEach(doc => {
+  const data = doc.data();
+  // console.log(`Document ${doc.ref.path}: ${data.type} unit: ${data.unit}`);
+  // Look for other documents with the same key
+  const key = data.key;
+  const keyQuery = documentCollection.where("key", "==", key);
+  const query = keyQuery.get().then(async snapshot => {
+    if (snapshot.size === 1) {
+      singles.add(key);
+    } else if (snapshot.size > 2) {
+      triplets.add(key);
+    } else if (snapshot.size === 2) {
+      if (!pairs.has(key)) {
+        pairs.add(key);
+        console.log(`Found ${snapshot.size} documents with key ${key}`);
+        console.log(`  ${snapshot.docs[0].ref.path} and ${snapshot.docs[1].ref.path}`);
+        let goodDoc, badDoc;
+        let badRef: admin.firestore.QueryDocumentSnapshot<admin.firestore.DocumentData>;
+        if (snapshot.docs[0].data().unit) {
+          goodDoc = snapshot.docs[0].data();
+          badDoc = snapshot.docs[1].data();
+          badRef = snapshot.docs[1];
+        } else if (snapshot.docs[1].data().unit) {
+          goodDoc = snapshot.docs[1].data();
+          badDoc = snapshot.docs[0].data();
+          badRef = snapshot.docs[0];
+        } else {
+          console.log(`   Neither document has metadata`);
+        }
+        if (goodDoc) {
+          // console.log("good doc", goodDoc);
+          // Make sure the main fields match in the two documents
+          let match = true;
+          should_match_fields.forEach(field => {
+            // Treat undefined and null as equivalent
+            const f1 = goodDoc[field] || null;
+            const f2 = badDoc[field] || null;
+            if (f1 !== f2) {
+              console.log(`      ${field} does not match: ${f1} != ${f2}`);
+              match = false;
+            }
+          });
+          // Make sure that the bad document has only nulls in all of the fields to copy
+          fields_to_copy.forEach(field => {
+            if (badDoc[field]) {
+              console.log(`      ${field} is not null: ${badDoc[field]}`);
+              match = false;
+            }
+          });
+          if (match) {
+            fixableDocsByType[goodDoc.type] += 1;
+            let logMsg = dryRun ? "    Would copy: " : "    Copy: ";
+            const updateDoc = { };
+            fields_to_copy.forEach(field => {
+              if (!dryRun) {
+                updateDoc[field] = goodDoc[field];
+              }
+              logMsg += ` ${field}: ${goodDoc[field]}  `;
+            });
+            if (dryRun) {
+              console.log(logMsg);
+            } else {
+              // Update the values in the bad document
+              await badRef.ref.set(updateDoc, { merge: true });
+              console.log(logMsg, "...done");
+            }
+          } else {
+            mismatchDocs.add(key);
+          }
+        }
+      }
+    }
+  });
+  promises.push(query);
+});
+
+await Promise.all(promises);
```

Comment:
If the number of documents found is large this approach could cause problems. All of the update code will be running in parallel which might exceed some limit in Firestore for number of parallel requests or number of requests per min.

I'd guess this issues would show up as exceptions through by `badRef.ref.set(updateDoc, { merge: true })`. I think that exception will be thrown up to this promise chain: `keyQuery.get().then(...`.  In the browser this will result in an unhandled exception in promise message. I'm not sure what node does with these. So to be safe you might want to wrap that `badRef.ref.set(updateDoc, { merge: true })` in a try catch so you can make sure to log any failures to update the documents.